app/ptxboa_functions.py

```diff
@@ -26,37 +28,50 @@ def calculate_results_single(api, settings):
     return res
 
 
-def calculate_results(api, settings):
-    # calculate results for all source regions:
-    results_list = []
-    for region in settings["region_list"]:
-        settings2 = settings.copy()
-        settings2["sel_region"] = region
-        result_single = pd.DataFrame.from_dict(
-            calculate_results_single(api, settings2)
-        ).reset_index(drop=True)
-        results_list.append(result_single)
-
-    results = pd.concat(results_list, ignore_index=True)[
-        ["source", "variable", "process_class", "value"]
-    ]
-
-    # aggregate costs (without LC):
-    res_costs = results.loc[results["variable"] != "LC"].pivot_table(
-        index="source", columns="process_class", values="value", aggfunc=sum
-    )
+def calculate_results(
+    api: PtxboaAPI, settings: dict, region_list: list = None
+) -> pd.DataFrame:
+    """Calculate results for all source regions."""
+    res_list = []
+
+    if region_list is None:
+        region_list = api.get_dimension("region")["region_name"]
+
+    for region in region_list:
+        res_single = api.calculate(
+            scenario=settings["sel_scenario"],
+            secproc_co2=settings["sel_secproc_co2"],
+            secproc_water=settings["sel_secproc_water"],
+            chain=settings["sel_chain"],
+            res_gen=settings["sel_res_gen_name"],
+            region=region,
+            country=settings["sel_country_name"],
+            transport=settings["sel_transport"],
+            ship_own_fuel=settings["sel_ship_own_fuel"],
+            output_unit=settings["selOutputUnit"],
+        )
+        res_list.append(res_single)
+    res = pd.concat(res_list)
+    return res
+
 
-    # Calculate total costs:
-    res_costs["Total"] = res_costs.sum(axis=1)
+def aggregate_costs(res_details: pd.DataFrame) -> pd.DataFrame:
+    """Aggregate detailed costs."""
+    # Exclude levelized costs:
+    res = res_details.loc[res_details["cost_type"] != "LC"]
+    res = res.pivot_table(
+        index="region", columns="process_type", values="values", aggfunc=sum
+    )
+    # calculate total costs:
+    res["Total"] = res.sum(axis=1)
 
-    # replace region codes with region names:
-    index_mapping = api.get_dimension("region").set_index("region_code")["region_name"]
-    res_costs.index = res_costs.index.map(index_mapping)
-    return res_costs
+    # TODO exclude countries with total costs of 0 - maybe remove later:
+    res = res.loc[res["Total"] != 0]
+    return res
```

Comment:
Why do you return a wide-format Table instead of long-format? Is this better for plotly?

Reply:
I changed to wide format to calculate the ``Total`` column and just kept it this way, its only two-dimensional data after all. But I am not strictly against long format, we can change this later if there is a good reason for it.